mqtt/mqtt-bridge/src/pump.rs

```diff
@@ -0,0 +1,283 @@
+
+        let (loader_shutdown, loader_shutdown_rx) = oneshot::channel::<()>();
+        let publish_handle = self.publish_handle.clone();
+        let persist = self.persist.clone();
+        let loader = self.loader.clone();
+        let mut client_shutdown = self.client_shutdown.clone();
+        let bridge_name = self.pump_context.bridge_name.clone();
+        let pump_type = self.pump_context.pump_type.clone();
+
+        // egress pump
+        let f1 = async move {
+            let mut loader_borrow = loader.borrow_mut();
+            let mut receive_fut = loader_shutdown_rx.into_stream();
+
+            debug!(
+                "{} bridge starting egress publication processing for {:?} pump...",
+                bridge_name, pump_type
+            );
+
+            loop {
+                let mut publish_handle = publish_handle.clone();
+                match select(receive_fut.next(), loader_borrow.try_next()).await {
+                    Either::Left((shutdown, _)) => {
+                        debug!("egress pump received shutdown signal");
+                        if let None = shutdown {
+                            error!(message = "unexpected behavior from shutdown signal while signalling bridge pump shutdown")
+                        }
+
+                        debug!("bridge pump stopped");
+                        break;
+                    }
+                    Either::Right((loaded_element, _)) => {
+                        debug!("egress pump extracted publication from store");
+
+                        if let Ok(Some((key, publication))) = loaded_element {
+                            debug!("publishing publication {:?} for egress pump", key);
+                            if let Err(e) = publish_handle.publish(publication).await {
+                                error!(message = "failed publishing publication for bridge pump", err = %e);
+                            }
+
+                            if let Err(e) = persist.remove(key) {
+                                error!(message = "failed removing publication from store", err = %e);
+                            }
+                        }
+                    }
+                }
+            }
+
+            debug!("pumps for {} bridge stopped...", bridge_name);
+        };
+
+        // incoming pump
+        let bridge_name = self.pump_context.bridge_name.clone();
+        let pump_type = self.pump_context.pump_type.clone();
+        let f2 = async move {
+            debug!(
+                "{} bridge starting ingress publication processing for pump {:?}...",
+                bridge_name, pump_type
+            );
+            self.client.handle_events().await;
+        };
+
+        let f1 = f1.fuse();
+        let f2 = f2.fuse();
+        let mut shutdown = shutdown.fuse();
+        pin_mut!(f1);
+        pin_mut!(f2);
+
+        select! {
+            _ = f1 => {
+                error!(message = "publish loop failed and exited for bridge pump");
```

Comment:
I believe this case is that the pump is stopped before shutdown, in which case it is an error right? See the shutdown in the select below on 317.

Reply:
if it finished on its own - yes, it is the error. but someone can request it to shut it down. in this case, it is a legit operation. In other words - if the result of operation Err - log the error, otherwise it is a clean shutdown

Reply:
I totally agree with this, but isn't that the way it is currently implemented?